### PR TITLE
Ranged capture

### PR DESCRIPTION
**Pull request by itanasi**
Adds ability for Ranged Units to Move onto unguarded Civilian Unit
If a unit moves on top of a Civilian Unit, execute a captureCivilian action (this should be free and thus consistent if also merging #5974 )
This still allows Ranged Units to Attack an unguarded Civilian unit. So it can be destroyed, all movement used, but doesn't move the Ranged Unit. That is, no change to Ranged Battle code.

### CODE REVIEW

**Review by xlenstra — APPROVED** [review]

**Comment by yairm210** [comment]
And what does the battle table display to the user when you attempt to do this?

**Comment by itanasi** [comment]
I should have included a screen shot.

You'll have two options. There's the Battle Tile dialog to attack (and thus kill the Civilian unit). Or you can move to the tile and capture the Civilian Unit as part of movement.  This is a bit inconsistent with melee units still doing the attack says Capture prompt. I could rework the whole system to be more consistent.

**Comment by itanasi** [comment]
image

Obviously this breaks the whole can't move through enemy unit though...

**Comment by itanasi** [comment]
Beware, this merge now also edits one of the tests. Highly recommend reviewing!

File: core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt [hunk 1]
```diff
@@ -619,6 +619,9 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
         if (!unit.canEnterForeignTerrain && !tile.canCivPassThrough(unit.civInfo)) return false
 
         val firstUnit = tile.getFirstUnit()
+        // allow Ranged Units to capture Civilian
+        if (firstUnit != null && firstUnit.isCivilian() && unit.isRanged() )
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt:623]: This is missing a check that the tile doesn't contain a military unit - don't count on getFirst to return the military unit if there is one

Comment (yairm210) [reply core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt:623]: ALSO
you don't check if the unit is off an ENEMY civ! You shouldn't be able to capture allied civilians with impunity

**Comment by itanasi** [comment]
Hm, didn't think about all of that. My bad.

Guess I should turn this into a revamp and move all civilian captures to movement. But also then add all the necessary checks and how it will trigger War as well.

File: core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt (hunk 1 above)
Comment (itanasi) [reply core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt:623]: > This is missing a check that the tile doesn't contain a military unit - don't count on getFirst to return the military unit if there is one

Why is that not guaranteed? Is this a multithreading thing? When I read getFirst() seemed it went in order.

Comment (yairm210) [reply core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt:623]: It is said that "programmers won't depends on what your library says it does, that will depend on observed results. Ordered results will be relief on to be ordered, and random ordering will be taken as a source of randomness".
Basically - if we want to be about to change the implementation of getFirst, we need to ensure that functions don't depend on a byproduct of its internal logic.

**Comment by itanasi** [comment]
I get the need for robustness and can make the change. But I guess I don't
understand then what the promised behavior is of getFirst() since the name
implies an ordering.

**Comment by will-ca** [comment]
> But I guess I don't understand then what the promised behavior is of getFirst() since the name implies an ordering.

Looks like it's meant for blocking movement through enemies. So in that case the promised/specified/relevant behaviour is that it will return a unit (belonging to the correct civInfo) if there are any units, or `null` if there are no units.

Compare to `.first()` and `.firstOrNull()` on unordered Kotlin `Iterable`s like `HashSet`. The "ordering" is "iteration order", but "iteration order" itself doesn't have any promises.

**Comment by itanasi** [comment]
Ah, thanks for the explanation! Makes sense.

Still new to Kotlin and it's conventions.

**Comment by yairm210** [comment]
> Hm, didn't think about all of that. My bad.
> 
> Guess I should turn this into a revamp and move all civilian captures to movement. But also then add all the necessary checks and how it will trigger War as well.

There's no need for that, we can easily have the tile move-through-able only if it contains a single, enemy, civilian unit, just by adding the 'enemy' check

**Comment by itanasi** [comment]
That would be functionally correct. But I'm starting to feel like there's a
lot of extra checks and exceptions. So want to see if unifying Civilian
Capture folded into the movement code (for both melee and ranged) would be
more robust.

Especially since we now also have Ranged Attack is 40hp and Civilian Cap
isn't an Attack. Basically fully close @ravignir original ticket. #3477

**Comment by itanasi** [comment]
Pretty sure the logic is correct now.
Running on an emulated save with a bunch of AIs in Modern Era, hard to tell how long canPassThrough takes since the AIs seem to take different tacts/choices.

File: core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt [hunk 2]
```diff
@@ -618,8 +619,14 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
         if (!unit.canEnterForeignTerrain && !tile.canCivPassThrough(unit.civInfo)) return false
 
         val firstUnit = tile.getFirstUnit()
-        if (firstUnit != null && firstUnit.civInfo != unit.civInfo && unit.civInfo.isAtWarWith(firstUnit.civInfo))
-            return false
+        if (firstUnit != null) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt:622]: I would put the `unit.civInfo != firstUnit.civInfo` check into the if as well

**Review by yairm210 — APPROVED** [review]
Definitely different, but this method allows the AI to both ignore capturable units and to do multiple things per turn, so well done!

**Review by xlenstra — APPROVED** [review]

**Pull request merged**